Add italian keycodes

giuseongit opened the pull request:
As mentioned in #69, here's the italian keycodes.

I do have a couple of questions:

1. CharToKeyCodeMap – Could you clarify the purpose of this variable?
1. I noticed that some keys are missing from the array. On an Italian keyboard, Alt Gr (the right Alt key) is used to type additional characters (like square [] and curly {} brackets, among others). Should those key codes be added to the array as well?

sago35 commented:
Thank you for creating PR.

> CharToKeyCodeMap – Could you clarify the purpose of this variable?

The `CharToKeyCodeMap` is used for converting `Text` in macros. For example, the `+` character is input as `Shift + =` on a US keyboard, but as `Shift + ;` on a Japanese keyboard. These conversion rules are documented in files such as `machine/usb/hid/keyboard.ascii`.  

Thus, when `1+2` is entered in `Text` within macros, if `CharToKeyCodeMap` does not exist, the default US keyboard conversion rules apply: `1` is input first, followed by `Shift + =`, and then `2`.  

However, when `CharToKeyCodeMap` is specified as `keycodes/japanese`, `Shift + ;` is used instead.  

Since `CharToKeyCodeMap` must be created based on ASCII codes, the `+` character should always be placed at position `43 (0x2b)`.

sago35 commented:
In `CharToKeyCodeMap`, the `altgrMask` needs to be added. It would be a good idea to add this in `keycodes/keycodes.go`.  

Reference:
https://github.com/tinygo-org/tinygo/blob/52983794d702af7a00833ae12b0d2e7175e46017/src/machine/usb/hid/keyboard/keycode.go#L233

Outside of `CharToKeyCodeMap`, `keycodes.KeyRightAlt` should be used. However, since it is not defined, it may need to be added. For now, you can try using `keycodes.KeyLeftAlt`, and if that doesn't work, you may need to define `KeyLeftAlt = TypeNormal | 0xE6`.

giuseongit commented:
Done, I added some commits in order to keep the logical steps clean.

Commit list:
1. 3bf903d - added AltGrMask
1. d82db8d - added italian keycodes
1. 3834c21 - fix shiftmask
1. 7c0c784 - added keyAltRight in base keycodes as well as italian and japanese keycodes

Note on commit no. 2: I saw that in the current implementation  ShiftMask has the value `0x0400`, however, in the [reference linked above](https://github.com/tinygo-org/tinygo/blob/52983794d702af7a00833ae12b0d2e7175e46017/src/machine/usb/hid/keyboard/keycode.go#L232) that mask has value of `0x0040`, so I fixed. If I'm wrong please let me know why those values differ and if I need to change the value of AltGrMask as well.

sago35 commented:
Please keep ShiftMask as 0x0400, and set AltGrMask to 0x0800.

If these values are 0x00FF or lower, there is a possibility that the mask could overwrite the values. I will take a closer look later, but there is also a chance that the settings in TinyGo's machine/usb/hid/keyboard are incorrect.

```go
package main

import (
	"fmt"
	"time"

	"github.com/sago35/tinygo-keyboard/keycodes"
	"github.com/sago35/tinygo-keyboard/keycodes/italian"
)

func main() {
	time.Sleep(3 * time.Second)
	fmt.Printf("KeyA               : %04X\n", italian.KeyA)
	fmt.Printf("KeyA   | ShiftMask : %04X | %04X : %04X\n", italian.KeyA, keycodes.ShiftMask, italian.KeyA|keycodes.ShiftMask)
	fmt.Printf("KeyF11             : %04X\n", italian.KeyF11)
	fmt.Printf("KeyF11 | ShiftMask : %04X | %04X : %04X\n", italian.KeyF11, keycodes.ShiftMask, italian.KeyF11|keycodes.ShiftMask)
}
```

```
$ tinygo flash --target xiao-rp2040 --size short --monitor .
   code    data     bss |   flash     ram
  59036    1924    4336 |   60960    6260
Connected to COM11. Press Ctrl-C to exit.
KeyA               : F004
KeyA   + ShiftMask : F004 + 0040 : F044
KeyF11             : F044
KeyF11 + ShiftMask : F044 + 0040 : F084
```

giuseongit commented:
I tried several values for AltGr mask (from 0x0800 to 0x4000) and non of them work. Any ideas on how to further inspect this issue?

sago35 commented:
@giuseongit Does the following work in your environment?  
I expect it to input `q` and `@` with your italian keyboard.
If it works, I can also implement AltGr support in my environment.

```go
package main

import (
	"machine/usb/hid"
	"machine/usb/hid/keyboard"
	"time"
)

func main() {
	_ = keyboard.Port()

	b1 := []byte{
		0x02,
		0x00, // no mod key
		0x02,
		0x14, // `q` key
		0x00, 0x00, 0x00, 0x00, 0x00}
	b2 := []byte{0x02, 0x00, 0x02, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00}
	b3 := []byte{
		0x02,
		0x40, // right alt key
		0x02,
		0x14, // `q` key
		0x00, 0x00, 0x00, 0x00, 0x00}

	// q
	time.Sleep(1000 * time.Millisecond)
	hid.SendUSBPacket(b1)
	time.Sleep(100 * time.Millisecond)
	hid.SendUSBPacket(b2)

	// altgr + q = @
	time.Sleep(1000 * time.Millisecond)
	hid.SendUSBPacket(b3)
	time.Sleep(100 * time.Millisecond)
	hid.SendUSBPacket(b2)
}
```

giuseongit commented:
@sago35 your code works as expected!
So I cleaned the PR , dropped the commit regarding the shift mask and set the altgr mask to 0x0800 (which don't work).
So the italian implementation is almost ready, apart from the AltGr implementation. Feel free to modify any commit of this pr as you wish, I'm willing to test it before merging.

sago35 commented:
@giuseongit  I have added handling for **AltGrMask**. Please check if it works as expected.  

```go
d.SetMacro(0, "@")
```

Also, please add definitions for **`KeyMacro0`** and similar keys.

sago35 commented on keycodes/italian/keycodes.go:
```diff
@@ -0,0 +1,310 @@
+	Key0 | keycodes.ShiftMask,           //  61   =
+	KeyPeriod | keycodes.ShiftMask,      //  62   >
+	KeyLessThan | keycodes.ShiftMask,    //  63   ?
+	KeyOGrave | keycodes.AltGrMask,      //  64   @
```
Isn't the correct definition here `KeyQ | AltGrMask`?

giuseongit replied:
Actually both are valid, but rout variant is less frequent among keyboards. The standard way of typing the `@` character is by pressing AltGr+Ograve. [https://www.tastiera-aggiornata.it/en/liv_principale.htm](here) you can find a reference on where to expect keys on an italian standard keyboard.

giuseongit commented:
Ok, it's working now.
I also added support for the combination `AltGr + Shift + [key]`, needed for keys `{` and `}`.
I've cleaned the history, leaving just two commits (in order to have a more streamlined history) and tested the macro with all keys as you pointed out, then I added the macro keycodes in the italian package. Give it a review, please.

sago35 reviewed (changes requested):
Thank you! The code is mostly good, but I would like one correction.  
The other parts were just questions.  

**`keyboard.go`** is likely incorrect, so please fix it.

giuseongit commented on keycodes/italian/keycodes.go:
```diff
@@ -0,0 +1,328 @@
+	KeyLeftAlt     = keycodes.TypeNormal | 0xE2
+	KeyWindows     = keycodes.TypeNormal | 0xE3
+	KeyRightCtrl   = keycodes.TypeNormal | 0xE4
+	KeyRightShift  = keycodes.TypeNormal | 0xE5
```
Oh no, I just forgot to add that line back between rebases

giuseongit commented on keyboard.go:
```diff
@@ -709,6 +709,17 @@ func (d *Device) RunMacro(no uint8) error {
+						k.Keyboard.Down(keycodes.KeyRightAlt)
+						k.Keyboard.Press(kc ^ keycodes.AltGrMask)
+						k.Keyboard.Up(keycodes.KeyRightAlt)
+						k.Keyboard.Up(keycodes.KeyLeftShift)
```
done
